### PR TITLE
Feature/new router

### PR DESCRIPTION
**Pull request by slawkens**
This PR allows to add new pages that are outside of system/pages directory (without touching the MyAAC structure), inside of plugins.

New dependency is added: nikic/fast-route. It's a very fast routing library installable through composer.

I also renamed/moved some pages, like createaccount into account/create so its more organized now.

Example of plugin routes, are in example-plugin.json

To Be Tested:
- [x] Account Management
- [x] Forum Actions
- [x] Custom Pages are working
- [x] Redirection is working
- [x] Admin Pages are not accessible

Known bugs:
- [x] News Archive by id, is bugged

More PRs like this are planned, to allow more options for plugins, and to allow easier update of MyAAC in the future.

### CODE REVIEW

File: admin/pages/accounts.php [hunk 1]
```diff
@@ -33,6 +33,11 @@
 }
 $web_acc = array("None", "Admin", "Super Admin", "(Admin + Super Admin)");
 $acc_type = array("None", "Normal", "Tutor", "Senior Tutor", "Gamemaster", "God");
+
+$groups = new OTS_Groups_List();
+if ($groups->getHighestId() == 6) {
```
Comment (gpedro) [inline admin/pages/accounts.php:38]: maybe move that to config file?

File: admin/pages/login.php [hunk 2]
```diff
@@ -10,9 +10,9 @@
 defined('MYAAC') or die('Direct access not allowed!');
 $title = 'Login';
 
-$twig->display('admin.login.html.twig', array(
-	'logout' => ($action == 'logout' ? 'You have  been logged out!'  : ''),
+$twig->display('admin.login.html.twig', [
+	'logout' => (ACTION == 'logout' ? 'You have  been logged out!'  : ''),
 	'account' => USE_ACCOUNT_NAME ? 'Name' : 'Number',
 	'account_login_by' => getAccountLoginByLabel(),
 	'errors' => isset($errors)? $errors : ''
```
Comment (gpedro) [inline admin/pages/login.php:17]:
```
'errors' => isset($errors)? $errors : ''
'errors' => $errors ?? ''
```

File: admin/pages/news.php [hunk 3]
```diff
@@ -118,7 +118,7 @@
 		'news_link_form' => '?p=news&action=' . ($action == 'edit' ? 'edit' : 'add'),
 		'news_id' => isset($id) ? $id : null,
 		'title' => isset($p_title) ? $p_title : '',
```
Comment (gpedro) [inline admin/pages/news.php:120]: ``
`
 'title' => isset($p_title) ? $p_title : '',
 'title' => $p_title ?? : '',
 ```

File: admin/pages/news.php [hunk 4]
```diff
@@ -118,7 +118,7 @@
...
-		'body' => isset($body) ? htmlentities($body, ENT_COMPAT, 'UTF-8') : '',
+		'body' => isset($body) ? escapeHtml($body) : '',
 		'type' => isset($type) ? $type : null,
 		'player' => isset($player) && $player->isLoaded() ? $player : null,
 		'player_id' => isset($player_id) ? $player_id : null,
```
Comment (gpedro) [inline admin/pages/news.php:124]:
```
 'type' => isset($type) ? $type : null,
 'type' => $type ?? null,
```

Comment (gpedro) [inline admin/pages/news.php:124]:
```
'player_id' => isset($player_id) ? $player_id : null,
'player_id' => $player_id ?? null,
```

File: system/login.php [hunk 5]
```diff
@@ -33,152 +28,131 @@
...
+				$tmp = array();
 
-			if(config('recaptcha_enabled') && !config('account_create_auto_login'))
-			{
-				require_once LIBS . 'GoogleReCAPTCHA.php';
-				if (!GoogleReCAPTCHA::verify('login')) {
-					$errors[] = GoogleReCAPTCHA::getErrorMessage();
-				}
+			$ip = $_SERVER['REMOTE_ADDR'];
+			$t = isset($tmp[$ip]) ? $tmp[$ip] : NULL;
```
Comment (gpedro) [inline system/login.php:59]:
```
 $t = isset($tmp[$ip]) ? $tmp[$ip] : NULL;
 $t = $tmp[$ip] ?? NULL;
```

File: system/pages/account/base.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+if(!$logged)
+{
+	if(!empty($errors))
+		$twig->display('error_box.html.twig', array('errors' => $errors));
+
+	$twig->display('account.login.html.twig', array(
+		'redirect' => isset($_REQUEST['redirect']) ? $_REQUEST['redirect'] : null,
+		'account' => USE_ACCOUNT_NAME ? 'Name' : 'Number',
+		'account_login_by' => getAccountLoginByLabel(),
+		'error' => isset($errors[0]) ? $errors[0] : null
```
Comment (gpedro) [inline system/pages/account/base.php:22]:
```
 'error' => $errors[0] ?"null
```

File: system/pages/account/change_info.php [hunk 7]
```diff
@@ -59,4 +69,4 @@
 		'account_country' => isset($account_country) ? $account_country : ''
```
Comment (gpedro) [inline system/pages/account/change_info.php:69]:
```
		'account_country' => isset($account_country) ? $account_country : ''
		'account_country' => $account_country ?? ''
```

File: system/pages/account/change_password.php [hunk 8]
```diff
@@ -10,6 +10,13 @@
...
+$title = 'Change Password';
+require __DIR__ . '/base.php';
+
+if(!$logged) {
+	return;
+}
+
 $new_password = isset($_POST['newpassword']) ? $_POST['newpassword'] : NULL;
 $new_password2 = isset($_POST['newpassword2']) ? $_POST['newpassword2'] : NULL;
 $old_password = isset($_POST['oldpassword']) ? $_POST['oldpassword'] : NULL;
```
Comment (gpedro) [inline system/pages/account/change_password.php:22]:
```
 $new_password = isset($_POST['newpassword']) ? $_POST['newpassword'] : NULL;
 $new_password2 = isset($_POST['newpassword2']) ? $_POST['newpassword2'] : NULL;
 $old_password = isset($_POST['oldpassword']) ? $_POST['oldpassword'] : NULL;

 $new_password = $_POST['newpassword'] ?? NULL;
 $new_password2 = $_POST['newpassword2'] ?? NULL;
 $old_password = $_POST['oldpassword'] ?? NULL;
```

File: system/pages/account/confirm_email.php [hunk 9]
```diff
@@ -11,7 +11,7 @@
 
 $title = 'Confirm Email';
 
-$hash = isset($_GET['v']) ? $_GET['v'] : '';
+$hash = isset($_GET['hash']) ? $_GET['hash'] : '';
```
Comment (gpedro) [inline system/pages/account/confirm_email.php:14]:
```
$hash = isset($_GET['hash']) ? $_GET['hash'] : '';
$hash = $_GET['hash'] ?? '';
```

File: plugins/example.json [hunk 10]
```diff
@@ -7,30 +7,37 @@
 	"contact": "nobody@example.org",
 	"require": {
 		"myaac": "0.4.3",
-		"myaac_": ">=0.7,<1.0", // support for defining versions like in composer (since 0.8)
+		"myaac_": ">=0.7,<1.0",
 		"php": "5.2.0",
-		"php_": ">5.4,<7.0", // support for defining versions like in composer (since 0.8)
+		"php_": ">5.4,<7.0",
```
[the remark this answers is not in the document]
Comment (slawkens) [reply plugins/example.json:12]: This is just an example :p myaac version defined there is also wrong

File: system/pages/news.php [hunk 11]
```diff
@@ -64,7 +64,6 @@
 				'icon' => $categories[$news['category']]['icon_id'],
 				'author' => $config['news_author'] ? $author : '',
 				'comments' => $news['comments'] != 0 ? getForumThreadLink($news['comments']) : null,
```
[the remark this answers is not in the document]
Comment (slawkens) [reply system/pages/news.php:66]: $news['comments'] is an id, thread id

File: system/routes.php [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * @package   MyAAC
+ * @author    Slawkens <slawkens@gmail.com>
+ * @copyright 2021 MyAAC
+ * @link      https://my-aac.org
+ */
+
+return [
+	['GET', '', '__redirect__/news'], // redirect empty URL to news
+	['GET', 'news/archive/{id:[0-9]+}[/]', 'news/archive.php'],
```
[the remark this answers is not in the document]
Comment (slawkens) [reply system/routes.php:13]: sure, that's a very clever idea

will definitely add it.

actually, you can define line this:
image

File: admin/pages/accounts.php (hunk 1 above)
Comment (slawkens) [reply admin/pages/accounts.php:38]: 👍

File: system/pages/changelog.php [hunk 13]
```diff
@@ -10,8 +10,8 @@
 defined('MYAAC') or die('Direct access not allowed!');
 $title = 'Changelog';
 
-$_page = isset($_GET['page']) ? $_GET['page'] : 0;
-$id = isset($_GET['id']) ? $_GET['id'] : 0;
+$_page = $_GET['page'] ?? 0;
+$id = $_GET['id'] ?? 0;
```
[the remark this answers is not in the document]
Comment (slawkens) [reply system/pages/changelog.php:14]: its escaped anyway, so i don't think it's necessary

when someone tried to spoof this, he will get 500 error

**Pull request merged**